Framework/Types/inc/MantidTypes/Core/DateAndTime.h

```diff
@@ -147,10 +147,66 @@ class MANTID_TYPES_DLL DateAndTime {
 private:
   /// A signed 64-bit int of the # of nanoseconds since Jan 1, 1990.
   int64_t _nanoseconds;
+
+  /// Max allowed nanoseconds in the time; 2^62-1
+  static constexpr int64_t MAX_NANOSECONDS = 4611686018427387903LL;
+
+  /// Min allowed nanoseconds in the time; -2^62+1
+  static constexpr int64_t MIN_NANOSECONDS = -4611686018427387903LL;
 };
 #pragma pack(pop)
+
+/** Default, empty constructor */
+inline DateAndTime::DateAndTime() : _nanoseconds(0) {}
+
+/** Construct a date from nanoseconds.
+ * @param total_nanoseconds :: nanoseconds since Jan 1, 1990 (our epoch).
+ */
+inline DateAndTime::DateAndTime(const int64_t total_nanoseconds) {
+  // Make sure that you cannot construct a date that is beyond the limits...
+  if (total_nanoseconds > MAX_NANOSECONDS)
+    _nanoseconds = MAX_NANOSECONDS;
+  else if (total_nanoseconds < MIN_NANOSECONDS)
+    _nanoseconds = MIN_NANOSECONDS;
+  else
+    _nanoseconds = total_nanoseconds;
+}
+
+/** + operator to add time.
+ * @param nanosec :: number of nanoseconds to add
+ * @return modified DateAndTime.
+ */
+inline DateAndTime DateAndTime::operator+(const int64_t nanosec) const {
+  return DateAndTime(_nanoseconds + nanosec);
+}
+
+/** + operator to add time.
+ * @param sec :: duration to add
+ * @return modified DateAndTime.
+ */
+inline DateAndTime DateAndTime::operator+(const double sec) const {
+  return this->operator+(nanosecondsFromSeconds(sec));
+}
+
+/** Nanoseconds from seconds, with limits
+ * @param sec :: duration in seconds, as a double
+ * @return int64 of the number of nanoseconds
+ */
+inline int64_t DateAndTime::nanosecondsFromSeconds(double sec) {
+  const double nano = sec * 1e9;
+  constexpr double minimum = static_cast<double>(MIN_NANOSECONDS);
+  constexpr double maximum = static_cast<double>(MAX_NANOSECONDS);
+  // Use these limits to avoid integer overflows
+  if (nano > maximum)
+    return MAX_NANOSECONDS;
+  else if (nano < minimum)
+    return MIN_NANOSECONDS;
+  else
+    return int64_t(nano);
+}
```

Comment:
Cannot use `return static_cast<int64_t>(boost::algorithm::clamp(name, min, max))` since casting to `double` and back to `int64_t` changes the value of min and max.